### PR TITLE
[M68k] Handle 16 bit MOVs to and from CCR 

### LINKED ISSUE #106210: [M68k] incorrect assertion on move from CC register to data register
The assertion "Need DR8 register to copy CCR" in M68kInstrInfo::copyPhysReg appears to be wrong. According to the Programmer's Reference Manual, MOVE from CCR takes a 32bit destination operand and zero extends the 5 bits in the CC register.

### PR DESCRIPTION
**Pull request by knickish**
Builds on @TechnoElf 's CCR MOV pr https://github.com/llvm/llvm-project/pull/107591 and adds some tests.

Fixes https://github.com/llvm/llvm-project/issues/106210.

### CODE REVIEW

**Comment by glaubitz** [comment]
CC @mshockwave @0x59616e

**Comment by glaubitz** [comment]
Thanks a lot for the contribution!

Could you start the subject with "[M68k]" and use a more descriptive title.

I'll have a go at testing your changes later today.

**Comment by knickish** [comment]
@glaubitz Thanks, much appreciated. I updated the title as requested. 

How are you testing these? I set up a [docker container chroot thing](https://github.com/knickish/m68k_tests) based on your guide on the debian forums, but it's pretty unwieldy to be honest

**Comment by glaubitz** [comment]
> @glaubitz Thanks, much appreciated. I updated the title as requested.
> 
> How are you testing these? I set up a [docker container chroot thing](https://github.com/knickish/m68k_tests) based on your guide on the debian forums, but it's pretty unwieldy to be honest

I'm currently doing cross-builds only as LLVM doesn't work natively on M68k at the moment due to the natural 2-byte alignment of the current ABI.

We're working on switching the ABI to 4-byte alignment by default, but there has been some resistance from a few people who insist on ABI compatibility.

However, both Debian and Gentoo agree that the port will only be maintainable in the future with 4-byte alignment by default.

Having said that, are there any specific issues you need help with?

**Comment by knickish** [comment]
I was just wondering what tests I should be running to verify that I'm not breaking anything, as the ones I have are super basic. The clang test suite maybe?

**Comment by glaubitz** [comment]
> I was just wondering what tests I should be running to verify that I'm not breaking anything, as the ones I have are super basic. The clang test suite maybe?

Yes, running the testsuite is fine. FWIW, we have a [buildbot for LLVM M68k](https://lab.llvm.org/buildbot/#/builders/27) to run integration tests.

File: llvm/lib/Target/M68k/M68kInstrInfo.cpp [hunk 1]
```diff
@@ -571,6 +573,17 @@ bool M68kInstrInfo::ExpandPUSH_POP(MachineInstrBuilder &MIB,
 }
 
 bool M68kInstrInfo::ExpandCCR(MachineInstrBuilder &MIB, bool IsToCCR) const {
+  switch (MIB->getOpcode()) {
+  case M68k::MOV8cd: {
+    // Promote used register to the next class
+    auto &Opd = MIB->getOperand(1);
```
Comment (mshockwave) [inline llvm/lib/Target/M68k/M68kInstrInfo.cpp:579]: please spell out the type here.

File: llvm/lib/Target/M68k/M68kInstrInfo.cpp [hunk 2]
```diff
@@ -571,6 +573,17 @@ bool M68kInstrInfo::ExpandPUSH_POP(MachineInstrBuilder &MIB,
 }
 
 bool M68kInstrInfo::ExpandCCR(MachineInstrBuilder &MIB, bool IsToCCR) const {
+  switch (MIB->getOpcode()) {
```
Comment (mshockwave) [inline llvm/lib/Target/M68k/M68kInstrInfo.cpp:576]: could we just use if statement here?

Comment (knickish) [reply llvm/lib/Target/M68k/M68kInstrInfo.cpp]: switched to an if statement

File: llvm/lib/Target/M68k/M68kInstrInfo.cpp (hunk 1 above)
Comment (knickish) [reply llvm/lib/Target/M68k/M68kInstrInfo.cpp]: done

File: llvm/lib/Target/M68k/M68kInstrInfo.cpp [hunk 3]
```diff
@@ -872,10 +889,10 @@ unsigned M68kInstrInfo::getGlobalBaseReg(MachineFunction *MF) const {
   //
   // NOTE
   // Normally M68k uses A5 register as global base pointer but this will
-  // create unnecessary spill if we use less then 4 registers in code; since A5
-  // is callee-save anyway we could try to allocate caller-save first and if
-  // lucky get one, otherwise it does not really matter which callee-save to
-  // use.
+  // create unnecessary spill if we use less then 4 registers in code; since
```
Comment (knickish) [inline llvm/lib/Target/M68k/M68kInstrInfo.cpp:892]: This got moved when I ran `clang-format` on this file, can remove the change if necessary

Comment (DianQK) [reply llvm/lib/Target/M68k/M68kInstrInfo.cpp:892]: You probably should run with `git clang-format HEAD~1`: [How to Submit a Patch](https://llvm.org/docs/Contributing.html#id9).

**Comment by knickish** [comment]
Ping @mshockwave, all PR comments have been addressed

**Review by mshockwave — APPROVED** [review]
LGTM thanks!

**Pull request merged**